engine/classes/Elgg/Security/Hmac.php

```diff
@@ -40,10 +40,8 @@ public function __construct($key, callable $comparator, $data, $algo = 'sha256')
 		if (!$data) {
 			throw new \InvalidArgumentException('$data cannot be empty');
 		}
-		if (is_array($data)) {
-			$data = json_encode(array_map('strval', $data));
-		} else {
-			$data = (string)$data;
+		if (!is_string($data)) {
+			$data = json_encode($data);
```

Comment:
I chose JSON because it would perhaps be less likely to change with PHP, but serialize would allow validating types of objects, tell the difference between a stdClass and an array.